test/e2e/deployment.go

```diff
@@ -374,17 +374,31 @@ func (d *Deployment) WaitForEnvoyUpdated() error {
 }
 
 func (d *Deployment) waitForEnvoyDaemonSetUpdated() error {
-	daemonSetUpdated := func() (bool, error) {
-		tempDS := new(apps_v1.DaemonSet)
-		if err := d.client.Get(context.TODO(), client.ObjectKeyFromObject(d.EnvoyDaemonSet), tempDS); err != nil {
+	contourPodImage := d.ContourDeployment.Spec.Template.Spec.Containers[0].Image
+	updatedPods := func() (bool, error) {
+		pods := new(v1.PodList)
+		labelSelectAppContour := &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(d.EnvoyDaemonSet.Spec.Selector.MatchLabels),
+			Namespace:     d.EnvoyDaemonSet.Namespace,
+		}
+		if err := d.client.List(context.TODO(), pods, labelSelectAppContour); err != nil {
 			return false, err
 		}
-		return tempDS.Status.ObservedGeneration == tempDS.Generation &&
-			tempDS.Status.NumberAvailable > 0 &&
-			tempDS.Status.NumberAvailable == tempDS.Status.DesiredNumberScheduled &&
-			tempDS.Status.UpdatedNumberScheduled == tempDS.Status.DesiredNumberScheduled, nil
+
+		updatedPods := 0
+		for _, pod := range pods.Items {
+			if pod.Spec.InitContainers[0].Image != contourPodImage {
+				continue
+			}
+			for _, cond := range pod.Status.Conditions {
+				if cond.Type == v1.PodReady && cond.Status == v1.ConditionTrue {
+					updatedPods++
+				}
+			}
+		}
+		return updatedPods == int(*&d.EnvoyDaemonSet.Status.DesiredNumberScheduled), nil
 	}
-	return wait.PollImmediate(time.Millisecond*50, time.Minute*3, daemonSetUpdated)
+	return wait.PollImmediate(time.Millisecond*50, time.Minute*3, updatedPods)
 }
 
 func (d *Deployment) waitForEnvoyDeploymentUpdated() error {
```

Comment:
I suppose it would make sense to use similar pod-level logic within this method too, for when we're testing envoy as a Deployment.